gcc/rust/typecheck/rust-tyty-rules.h

```diff
@@ -1069,12 +1107,66 @@ class StrRules : public BaseRules
 
   void visit (StrType &type) override { resolved = type.clone (); }
 
+  void visit (NeverType &type) override { resolved = base->clone (); }
+
 private:
   BaseType *get_base () override { return base; }
 
   StrType *base;
 };
 
+class NeverRules : public BaseRules
+{
+  using Rust::TyTy::BaseRules::visit;
+
+public:
+  NeverRules (NeverType *base) : BaseRules (base), base (base) {}
+
+  virtual void visit (TupleType &type) override { resolved = type.clone (); }
```

Comment:
I would have assumed that the only methods you would implement for a never type would be InferType and NeverType as only those types can be unified. This class will allow ints to unify with never types and coerce ints to be !.

Reply:
Yes, but actually this problem does not exist now. It is because `!` has not been integrated and code like `let a: !` or `fn foo() -> ! { /* ... */ }` simply does not compile.

In the future, when we integrate the `!` type, I think we need to explicitly distinguish `unify` and `assign`. For example, currently both of the following two code segments will unify `!` and `<integer>`, but the first is legal and the second is illegal.
```rust
let a = if x > y {
    return -1;
} else {
    y
};
```
```rust
let a: !;
a = 1;
```

Reply:
And what's more, assigning `<integer>` to `!` is illegal and assigning `!` to `<integer>` is legal. At present, it seems that we do not distinguish between `A->unify(B)` and `B->unify(A)`.

Reply:
I see what you mean since the ! is not lexically in scope as a builtin type yet. Thats true.

The other problem such as your 2nd example

```
let a: !;
a = 1;
```

The way your unify rules are in this PR will allow this to work. In order to fix it all you should have to do is remove the visitors in NeverRules to simply be

```
class NeverRules {
  virtual void visit (InferType &type) override { if type.get_infer_kind() == InferTypeKind::GENERAL { resolved = type.clone() } }
  
  virtual void visit (NeverType &type) override { resolved = type.clone (); }
}
```

When you dont implement the other visitors it means they defaulty back to a type mis match error for you. At the moment your unify rules will let that be valid.

Reply:
I think for the ordering problem, that is up to the caller to make sure it is done in order i don't think there is a general way to enforce that at present.

Reply:
Oh wait i just reread your comment is <integer> unified vis ! never actually legal? Ignore my review comments if this is true.

Reply:
Ok i just tried this: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0d0d292b1139de6201392255007c0691

It shows that assigning integer to ! is illegal. So in this case my comments still stand all you have to do is a tweak to your rules here.

Reply:
> When you dont implement the other visitors it means they defaulty back to a type mis match error for you. At the moment your unify rules will let that be valid.

Then won't the following code generate a type-mismatched error?
```rust
    let b = if c > 0 {
        return 1; // ! type
    } else {
        c // <integer> type
    };
```
```rust
    let b = if c > 0 {
        c // <integer> type
    } else {
        return -1; // ! type
    };
```
Here we need to unify `!` with any other types.

Reply:
> Oh wait i just reread your comment is integer unified against never actually legal in rust? Ignore my review comments if this is true. Thanks
Here is the code I've tested with rustc.
```rust
let a: !;
a = 1; // It does NOT compile.
```
```rust
let a: u32;
a = return -1; // It DOES compile.
```

Reply:
Interesting, maybe rust playgorund is not right but i was trying this:

```
#![feature(never_type)]
fn main() {
    let x: ! = {
        return 123
    };
}
```

This gives:

```
error[E0308]: mismatched types
 --> src/main.rs:5:12
  |
5 |     return 123
  |            ^^^ expected `()`, found integer
  ```

Reply:
Sorry for being pedantic here its looking likely your code is correct here but lets be explicit about what should work or shouldn't.

Reply:
Whats interesting it that the documentation says the example i ran should have worked as you said:

https://doc.rust-lang.org/std/primitive.never.html

Reply:
Oh actually, it expected `()` simply because the `main` function's return type is `()`.